fedimint-atomic-broadcast/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+/// shutdown sender or the [OrderedItem] receiver is dropped.
+pub use broadcast::run;
+
+/// This keychain implements naive threshold schnorr signatures over secp256k1.
+/// The broadcasts uses this keychain to sign messages for other nodes and create
+/// the threshold signatures for the signed blocks.
+pub use keychain::Keychain;
+
+/// The majority of these messages need to be delivered to the intended [Recipient]
+/// in order for the broadcast to make progress. However, the broadcast does not assume
+/// a reliable network layer and implements all necessary retry logic. Therefore, the
+/// caller just has to try to send the message once and can otherwise discard it immediatly.
+#[derive(Clone,Debug, Encodable, Decodable)]
+pub enum Message {
+    NetworkData(Vec<u8>),
+    BlockRequest(u64),
+    Block(SignedBlock),
+}
+
+/// This enum defines the intented destination of a [Message].
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Recipient {
+    Everyone,
+    Peer(PeerId),
+}
+
+/// This enum specifies wether an [OrderedItem] has been accepted or discarded
+/// by the Fedimint Consensus.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Decision {
+    Accept,
+    Discard,
+}
+
+/// If two correct nodes obtain two ordered items from the broadcast they
+/// are guaranteed to be in the same order. However, an ordered items is
+/// only guaranteed to be seen by all correct nodes if and only if a correct
+/// node decidecs to accept and item.
+#[derive(Debug)]
+pub struct OrderedItem {
+    pub item: Vec<u8>,
+    pub block_index: u64,
+    pub peer_id: PeerId,
+    pub decision_sender: oneshot::Sender<Decision>,
+}
+
+/// An accepted item is created and recorded in the corresponding block if and only
+/// if  Fedimint Consensus decides to accept an [OrderedItem].
+#[derive(Clone, Debug, PartialEq, Eq, Encodable, Decodable)]
+pub struct AcceptedItem {
+    pub item: Vec<u8>,
+    pub peer_id: PeerId,
+}
+
+/// All items ordered in a session that have been accepted by Fedimint Consensus
+/// are recorded in the corresponding block. A running Federation produces a [Block]
+/// roughly every five minutes.  Therefore, just like in Bitcoin, a [Block]
+/// might be empty if no items are ordered in that time or all ordered items
+/// are discarded by Fedimint Consensus.
+#[derive(Clone, Debug, PartialEq, Eq, Encodable, Decodable)]
+pub struct Block {
+    pub index: u64,
+    pub items: Vec<AcceptedItem>,
+}
+
+impl Block {
+    /// A blocks header consists of 40 bytes formed by its index in big endian bytes
+    /// concatenated with the merkle root build from the consensus hashes of its [AcceptedItem]s 
+    /// or 32 zero bytes if the block is empty. The use of a merkle tree allows for 
+    /// efficient inclusion proofs of accepted consensus items for clients.
+    pub fn header(&self) -> [u8; 40] {
+        let mut header = [0; 40];
+
+        header[..8].copy_from_slice(&self.index.to_be_bytes());
```

Comment:
A different way to go about writing to that mutable array would be creating a `Cursor` that operates on it. [Example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0edb2a213afe40520ffad87a5d80404d)